rasa/core/training/story_reader/yaml_story_reader.py

```diff
@@ -25,154 +28,202 @@
-        try:
-            yaml_content = rasa.utils.io.read_yaml_file(filename)
-            if not isinstance(yaml_content, dict):
-                common_utils.raise_warning(
-                    f"Failed to read '{filename}'. It should be a Yaml dict."
-                )
-                return []
-
-            stories = yaml_content.get(KEY_STORIES)  # pytype: disable=attribute-error
-            if not stories:
-                return []
+        """Read stories or rules from file.
 
-            self._parse_stories(stories)
-            self._add_current_stories_to_result()
-            return self.story_steps
+        Args:
+            filename: Path to the story/rule file.
 
+        Returns:
+            `StoryStep`s read from `filename`.
+        """
+        try:
+            yaml_content = rasa.utils.io.read_yaml_file(filename)
         except ValueError as e:
             common_utils.raise_warning(
                 f"Failed to read YAML from '{filename}', it will be skipped. Error: {e}"
             )
+            return []
 
-        return []
+        if not isinstance(yaml_content, dict):
+            common_utils.raise_warning(
+                f"Failed to read '{filename}'. It should be a YAML dictionary."
+            )
+            return []
 
-    def _parse_stories(self, stories: List[Dict[Text, Any]]) -> None:
-        for story_item in stories:
-            if not isinstance(story_item, dict):
+        stories = yaml_content.get(KEY_STORIES, [])
+        self._parse_data(stories, is_rule_data=False)
+
+        rules = yaml_content.get(KEY_RULES, [])
+        self._parse_data(rules, is_rule_data=True)
+
+        self._add_current_stories_to_result()
+
+        return self.story_steps
+
+    def _parse_data(self, data: List[Dict[Text, Any]], is_rule_data: bool) -> None:
+        item_title = self._get_item_title(is_rule_data)
+
+        for item in data:
+            if not isinstance(item, dict):
                 common_utils.raise_warning(
-                    f"Unexpected block found in '{self.source_name}': \n"
-                    f"{story_item}\n"
-                    f"Items under the '{KEY_STORIES}' key must be YAML dictionaries. "
-                    f"It will be skipped.",
-                    docs=DOCS_URL_STORIES,
+                    f"Unexpected block found in '{self.source_name}':\n"
+                    f"{item}\nItems under the "
+                    f"'{self._get_plural_item_title(is_rule_data)}' key must be YAML "
+                    f"dictionaries. It will be skipped.",
+                    docs=self._get_docs_link(is_rule_data),
                 )
                 continue
 
-            if KEY_STORY_NAME in story_item.keys():
-                self._parse_plain_story(story_item)
+            if item_title in item.keys():
+                self._parse_plain_item(item, is_rule_data)
+
+    def _parse_plain_item(self, item: Dict[Text, Any], is_rule_data: bool) -> None:
```

Comment:
do we actually need to pass through the `is_rule_data` or can we take it from the current `step_builder`?

Reply:
optional: To be honest, I'm not a huge fan of the [flag argument](https://martinfowler.com/bliki/FlagArgument.html#:~:text=A%20flag%20argument%20is%20a,do%20this%3A%20regular%20and%20premium%20.). It works at this stage (that's why `optional` but makes things 
- complex: additional parameter everywhere + if else -> I think this one could also be solved by moving the flag to the StoryStepBuilder
- inflexible: It currently only makes sense because Rules and Stories are very similar. If you want to deal with unique features of one (e.g. currently you could mix a `story` into `rules` section  or in the future you might want to forbid e2e usage in rules). In that case I think it might make sense to have two classes which inherit from a common base class and can just override certain functionality (ala https://refactoring.guru/replace-conditional-with-polymorphism) . 

On the other hand we are very early stage it might make more sense to find a suited abstraction once the dust settled. Actually, as I'm writing this I'm slightly favoring in waiting 😄 However, still wanted to share my thoughts.

Reply:
In this case the flag mainly determines what docstring is used and what warning messages are emitted. IMO this doesn't justify two sets of methods, as it would make the class unnecessarily bloated

Reply:
it would just be having one class with the common behavior and having two subclasses which overwrite the docstring content / warning messages, no? 

What do we do about the fact that you can currently mix a `story` step in a `rules` section and vice versa?

Reply:
but we'd be overriding every `_parse_<X>` method?

> What do we do about the fact that you can currently mix a story step in a rules section and vice versa?

this is not an issue right? if we're in the `stories` section and encounter a `- rule:`, it will just be read as a rule

Reply:
> but we'd be overriding every _parse_<X> method?

Nope, as long as we use something like https://en.wikipedia.org/wiki/Template_method_pattern

> this is not an issue right? if we're in the stories section and encounter a - rule:, it will just be read as a rule

That's not ideal imo. But it should probably be taken care of by yaml schema validation. 

Let's keep it like that for now and see how things evolve 👍